drivers/rtc/rtc_nxp_irtc.c

```diff
@@ -0,0 +1,502 @@
+			((irtc_reg->CTRL & RTC_CTRL_DST_EN_MASK) >> RTC_CTRL_DST_EN_SHIFT);
+	}
+
+	/* There is no nano second support for IRTC */
+	timeptr->tm_nsec = 0;
+	/* There is no day of the year support for IRTC */
+	timeptr->tm_yday = -1;
+
+	return 0;
+}
+
+#if defined(CONFIG_RTC_ALARM) || defined(__DOXYGEN__)
+static int nxp_irtc_alarm_get_supported_fields(const struct device *dev, uint16_t id,
+					       uint16_t *mask)
+{
+	ARG_UNUSED(dev);
+
+	if (id != 0) {
+		return -EINVAL;
+	}
+
+	*mask = (RTC_ALARM_TIME_MASK_SECOND | RTC_ALARM_TIME_MASK_MINUTE |
+		 RTC_ALARM_TIME_MASK_HOUR | RTC_ALARM_TIME_MASK_MONTHDAY |
+		 RTC_ALARM_TIME_MASK_MONTH | RTC_ALARM_TIME_MASK_YEAR);
+
+	return 0;
+}
+
+static int nxp_irtc_alarm_set_time(const struct device *dev, uint16_t id, uint16_t mask,
+				   const struct rtc_time *timeptr)
+{
+	const struct nxp_irtc_config *config = dev->config;
+	struct nxp_irtc_data *data = dev->data;
+	RTC_Type *irtc_reg = config->base;
+
+	if (id != 0 || (mask && (timeptr == 0)) ||
+	    (timeptr && !is_rtc_time_in_bounds(timeptr, mask))) {
+		return -EINVAL;
+	}
+
+	uint32_t key = irq_lock();
+
+	nxp_irtc_unlock_registers(irtc_reg);
+
+	if ((mask & RTC_ALARM_TIME_MASK_SECOND) == RTC_ALARM_TIME_MASK_SECOND) {
+		irtc_reg->ALM_SECONDS = RTC_ALM_SECONDS_ALM_SEC(timeptr->tm_sec);
+	}
+
+	if ((mask & RTC_ALARM_TIME_MASK_MINUTE) == RTC_ALARM_TIME_MASK_MINUTE) {
+		irtc_reg->ALM_HOURMIN = RTC_ALM_HOURMIN_ALM_MIN(timeptr->tm_min);
+	}
+
+	if ((mask & RTC_ALARM_TIME_MASK_HOUR) == RTC_ALARM_TIME_MASK_HOUR) {
+		irtc_reg->ALM_HOURMIN |= RTC_ALM_HOURMIN_ALM_HOUR(timeptr->tm_hour);
+	}
+
+	if ((mask & RTC_ALARM_TIME_MASK_MONTHDAY) == RTC_ALARM_TIME_MASK_MONTHDAY) {
+		irtc_reg->ALM_DAYS = RTC_ALM_DAYS_ALM_DAY(timeptr->tm_mday);
+	}
+
+	if ((mask & RTC_ALARM_TIME_MASK_MONTH) == RTC_ALARM_TIME_MASK_MONTH) {
+		irtc_reg->ALM_YEARMON = RTC_ALM_YEARMON_ALM_MON(timeptr->tm_mon + 1);
+	}
+
+	if ((mask & RTC_ALARM_TIME_MASK_YEAR) == RTC_ALARM_TIME_MASK_YEAR) {
+		irtc_reg->ALM_YEARMON |=
+			RTC_ALM_YEARMON_ALM_YEAR(timeptr->tm_year - RTC_NXP_IRTC_YEAR_OFFSET);
+	}
+
+	/* Clearing out the ALARM Flag Field then setting the correct value */
+	irtc_reg->CTRL &= ~(0xC);
+	switch (mask) {
+	case 0x0F: {
+		irtc_reg->CTRL |= RTC_CTRL_ALM_MATCH(0x4);
+	} break;
+	case 0x1F: {
+		irtc_reg->CTRL |= RTC_CTRL_ALM_MATCH(0x8);
+	} break;
+	case 0x3F: {
+		irtc_reg->CTRL |= RTC_CTRL_ALM_MATCH(0xC);
+	} break;
+	default: {
+		irtc_reg->CTRL |= RTC_CTRL_ALM_MATCH(0x0);
+	}
+	}
+
+	irtc_reg->IER |= 0x4;
+	data->alarm_mask = mask;
+	data->is_alarm_pending = 1;
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+static int nxp_irtc_alarm_get_time(const struct device *dev, uint16_t id, uint16_t *mask,
+				   struct rtc_time *timeptr)
+{
+	const struct nxp_irtc_config *config = dev->config;
+	struct nxp_irtc_data *data = dev->data;
+	RTC_Type *irtc_reg = config->base;
+	uint16_t curr_alarm_mask = data->alarm_mask;
+
+	if (id != 0 || !timeptr) {
+		return -EINVAL;
+	}
+
+	*mask = 0;
```

Comment:
might be better to make a stack variable before writing back once to this, what do you think?